tests/test_build.py

```diff
@@ -64,36 +50,40 @@ def test_build_all():
 
                    %(test_name)s/%(test_name)s
                 """ % {'test_name': test_name})
-        )
+                              )
+    return srcdir
 
-    with mock.patch('sphinx.builders.linkcheck.requests') as requests:
-        requests.head = request_session_head
 
+@pytest.mark.parametrize(
+    "buildername",
+    [
         # note: no 'html' - if it's ok with dirhtml it's ok with html
-        for buildername in ['dirhtml', 'singlehtml', 'latex', 'texinfo', 'pickle',
-                            'json', 'text', 'htmlhelp', 'qthelp', 'epub2', 'epub',
-                            'applehelp', 'changes', 'xml', 'pseudoxml', 'man',
-                            'linkcheck']:
-            yield verify_build, buildername, srcdir
+        'dirhtml', 'singlehtml', 'latex', 'texinfo', 'pickle', 'json', 'text',
+        'htmlhelp', 'qthelp', 'epub2', 'epub', 'applehelp', 'changes', 'xml',
+        'pseudoxml', 'man', 'linkcheck',
+    ],
+)
+def test_build_all(make_app, nonascii_srcdir, buildername):
+    with mock.patch('sphinx.builders.linkcheck.requests') as requests:
+        requests.head = request_session_head
+        app = make_app(buildername, srcdir=nonascii_srcdir)
+        # if listdir is not empty, we can use built cache
+        if not app.outdir.listdir():
+            app.build()
 
 
-@with_tempdir
-def test_master_doc_not_found(tmpdir):
+def test_master_doc_not_found(tmpdir, make_app):
+    tmpdir = path(tmpdir)
```

Comment:
How about making `tempdir` fixture instead `tmpdir`? Wrapping everytime are very boring for me.